Date: 23/08/19

The dataset used for NN is captured in the lab environment. The appliances used are Acer Laptop Charger and Electric Fan

The data below are as indicated.

Status

00 0 Off State

01 1 Fan On

10 2 Laptop On

11 3 Laptop + Fan

Initialize weights wij with arbitrary values

Repeat
        
        Pick a training example <x,oˇ> (x is the input, oˇ is the expected output)
        Compute the sum for each neuron: Sj=∑i=1Nwijxi
        Compute outputs (o) of the network: oj=f(S)
        For each output, compute the error: δj=(oˇj−oj)
        Update each synaptic weights of the network: wij=wij+η.δj.xi.df(S)dS
Until the training set is empty

Date: 28/08/19
Changes made from previous version 1.2
1. Reshape Y data and normalize it properly
2. Retrain the model and extract the prediction output into csv to visualize

a. The output with the respective model is still wrong refer to the csv file for reference

In [1]:
import pandas as pd
import numpy as np
df = pd.read_csv(r'C:\Users\shich\Desktop\FYP\Documentation\Data Acquisition\Lab\Profile#1.csv')#Reading the house price data
df.describe()

Aggregate Power  Aggregate Current  Calculated Aggregated Current  \
count     10613.000000       10613.000000                   10613.000000   
mean         80.373222           0.349449                       0.400326   
std          42.941137           0.186701                       0.188029   
min           0.000000           0.000000                       0.000000   
25%          52.900000           0.230000                       0.240000   
50%          96.600000           0.420000                       0.500000   
75%         112.700000           0.490000                       0.530000   
max         161.000000           0.700000                       0.770000   

       Calculated Aggregated Power           Fan        Laptop        Status  
count                 10613.000000  10613.000000  10613.000000  10613.000000  
mean                     92.074984      0.164479      0.235847      2.560539  
std                      43.246625      0.098429      0.139470      0.889290  
min                       0.000000      0.000000      0.000000      0.000000  
25%                      55.200000      0.000000      0.200000      3.000000  
50%                     115.000000      0.220000      0.290000      3.000000  
75%                     121.900000      0.230000      0.320000      3.000000  
max                     177.100000      0.280000      0.540000      3.000000

In [2]:
#converting in the arrays so that our machine could process
dataset = df.values
dataset

array([[1.61e+01, 7.00e-02, 0.00e+00, ..., 0.00e+00, 0.00e+00, 0.00e+00],
       [0.00e+00, 0.00e+00, 0.00e+00, ..., 0.00e+00, 0.00e+00, 0.00e+00],
       [2.30e+00, 1.00e-02, 0.00e+00, ..., 0.00e+00, 0.00e+00, 0.00e+00],
       ...,
       [0.00e+00, 0.00e+00, 0.00e+00, ..., 0.00e+00, 0.00e+00, 0.00e+00],
       [0.00e+00, 0.00e+00, 0.00e+00, ..., 0.00e+00, 0.00e+00, 0.00e+00],
       [0.00e+00, 0.00e+00, 0.00e+00, ..., 0.00e+00, 0.00e+00, 0.00e+00]])

As we can see the our values are stored in the array now.

Now we will split our dataset into input features (X) and the feature we wish to predict (Y).

In [3]:
X = dataset[:,3:6]

In [4]:
Y= dataset[:,6:7]
Y.shape

(10613, 1)

In [5]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler(feature_range=(0,1))
Y_shape = scaler.fit_transform(Y)
#Y_series = pd.Series(Y)
Y_df = pd.DataFrame(Y_shape)
#This is to see the full result of after conversion
#Y_df.to_csv('new.csv')
Y_df.shape

(10613, 1)

In [6]:
#Ensuring that scaling of our input features are similar with the help of sckit-learn
from sklearn import preprocessing

#using the function 'min-max scaler' which scales the dataset so that
#all the features lie between 0 and 1
min_max_scaler = preprocessing.MinMaxScaler()
X_scale = min_max_scaler.fit_transform(X)
X_scale

array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       ...,
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])

In [7]:
#We will now split our data into the training set and test set with the help of the sklearn
from sklearn.model_selection import train_test_split
X_train, X_val_and_test, Y_train, Y_val_and_test = train_test_split(X_scale, Y_df, test_size=0.3)
#This tell sklearn that your_val_and_test size will be 30% of the overall dataset.

#using the same function to do the split again on val_and_test
X_val, X_test, Y_val, Y_test = train_test_split(X_val_and_test, Y_val_and_test, test_size=0.5)

print(X_train.shape, X_val.shape, X_test.shape, Y_train.shape, Y_val.shape, Y_test.shape)

(7429, 3) (1592, 3) (1592, 3) (7429, 1) (1592, 1) (1592, 1)


### As we can see, We have 7429 data points while the validation and test set has 1592 data points each.
#### Till now we have done following things step-wise
1. Read in the CSV file
2. Split our dataset into the input features and the label
3. Scale the data so that the input features have similar orders of the magnitude.
4. Splitting our dataset into the training set, the validation set and the test set.

# Building And Training The Neural Network

In [8]:
from keras.models import Sequential
from keras.layers import Dense

#Visualization purpose
from keras.utils import plot_model

Using TensorFlow backend.


Since we want to have have follwing kind of layers in the Neural Network
1. Hidden Layer 1:32 neurons, ReLU activation
2. Hidden Layer 2:32 neurons, ReLU activation
3. Output Layer 1 neuron, Sigmoid activation

We will run the follwoing code to fulfill our requirements

In [9]:
#This says that we will store our model in the variable 'model' and we will
#describe it sequentially(layer by layer)
model = Sequential([
    Dense(32, activation='relu', input_shape=(3,)),
    Dense(32, activation='relu'),
    Dense(1, activation='sigmoid'),
])
plot_model(model, to_file='model_plot.png', show_shapes=True, show_layer_names=True)
model.summary()


W0828 20:53:48.706621 21960 deprecation_wrapper.py:119] From C:\ProgramData\Anaconda3\lib\site-packages\keras\backend\tensorflow_backend.py:74: The name tf.get_default_graph is deprecated. Please use tf.compat.v1.get_default_graph instead.

W0828 20:53:48.723034 21960 deprecation_wrapper.py:119] From C:\ProgramData\Anaconda3\lib\site-packages\keras\backend\tensorflow_backend.py:517: The name tf.placeholder is deprecated. Please use tf.compat.v1.placeholder instead.

W0828 20:53:48.725190 21960 deprecation_wrapper.py:119] From C:\ProgramData\Anaconda3\lib\site-packages\keras\backend\tensorflow_backend.py:4138: The name tf.random_uniform is deprecated. Please use tf.random.uniform instead.



_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_1 (Dense)              (None, 32)                128       
_________________________________________________________________
dense_2 (Dense)              (None, 32)                1056      
_________________________________________________________________
dense_3 (Dense)              (None, 1)                 33        
Total params: 1,217
Trainable params: 1,217
Non-trainable params: 0
_________________________________________________________________


### Filling in the best numbers


In [10]:
#model.compile(optimizer='sgd',
#             loss='binary_crossentropy',
#             metrics=['accuracy'])
#sgd means stochastic gradient descent
#loss function for outputs that 
#metrics=accuracy means we want to track accuracy on top of the loss function
model.compile(optimizer='sgd',
             loss='binary_crossentropy',
             metrics=['accuracy'])
model.summary()

W0828 20:53:49.114609 21960 deprecation_wrapper.py:119] From C:\ProgramData\Anaconda3\lib\site-packages\keras\optimizers.py:790: The name tf.train.Optimizer is deprecated. Please use tf.compat.v1.train.Optimizer instead.

W0828 20:53:49.163618 21960 deprecation_wrapper.py:119] From C:\ProgramData\Anaconda3\lib\site-packages\keras\backend\tensorflow_backend.py:3376: The name tf.log is deprecated. Please use tf.math.log instead.

W0828 20:53:49.171199 21960 deprecation.py:323] From C:\ProgramData\Anaconda3\lib\site-packages\tensorflow\python\ops\nn_impl.py:180: add_dispatch_support.<locals>.wrapper (from tensorflow.python.ops.array_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where


_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_1 (Dense)              (None, 32)                128       
_________________________________________________________________
dense_2 (Dense)              (None, 32)                1056      
_________________________________________________________________
dense_3 (Dense)              (None, 1)                 33        
Total params: 1,217
Trainable params: 1,217
Non-trainable params: 0
_________________________________________________________________


In [11]:
#By the function fit we are fitting the parameters to the data
hist = model.fit(X_train, Y_train,
                batch_size=32, epochs=100,
                validation_data = (X_val, Y_val))

W0828 20:53:49.433467 21960 deprecation_wrapper.py:119] From C:\ProgramData\Anaconda3\lib\site-packages\keras\backend\tensorflow_backend.py:986: The name tf.assign_add is deprecated. Please use tf.compat.v1.assign_add instead.



Train on 7429 samples, validate on 1592 samples
Epoch 1/100
7429/7429 [==============================] - 1s 81us/step - loss: 0.4352 - acc: 0.7586 - val_loss: 0.3271 - val_acc: 0.7707
Epoch 2/100
7429/7429 [==============================] - 0s 43us/step - loss: 0.3224 - acc: 0.7582 - val_loss: 0.2963 - val_acc: 0.7707
Epoch 3/100
7429/7429 [==============================] - 0s 48us/step - loss: 0.2988 - acc: 0.7582 - val_loss: 0.2758 - val_acc: 0.7707
Epoch 4/100
7429/7429 [==============================] - 0s 48us/step - loss: 0.2783 - acc: 0.7582 - val_loss: 0.2583 - val_acc: 0.7707
Epoch 5/100
7429/7429 [==============================] - 0s 45us/step - loss: 0.2635 - acc: 0.7582 - val_loss: 0.2479 - val_acc: 0.7707
Epoch 6/100
7429/7429 [==============================] - 0s 46us/step - loss: 0.2553 - acc: 0.7770 - val_loss: 0.2418 - val_acc: 0.8260
Epoch 7/100
7429/7429 [==============================] - 0s 43us/step - loss: 0.2501 - acc: 0.8203 - val_loss: 0.2370 - val_acc: 0.8260


7429/7429 [==============================] - 0s 45us/step - loss: 0.2158 - acc: 0.8203 - val_loss: 0.2038 - val_acc: 0.8260
Epoch 61/100
7429/7429 [==============================] - 0s 47us/step - loss: 0.2157 - acc: 0.8203 - val_loss: 0.2038 - val_acc: 0.8260
Epoch 62/100
7429/7429 [==============================] - 0s 49us/step - loss: 0.2157 - acc: 0.8203 - val_loss: 0.2038 - val_acc: 0.8260
Epoch 63/100
7429/7429 [==============================] - 0s 47us/step - loss: 0.2156 - acc: 0.8203 - val_loss: 0.2040 - val_acc: 0.8260
Epoch 64/100
7429/7429 [==============================] - 0s 47us/step - loss: 0.2157 - acc: 0.8203 - val_loss: 0.2037 - val_acc: 0.8260
Epoch 65/100
7429/7429 [==============================] - 0s 46us/step - loss: 0.2156 - acc: 0.8203 - val_loss: 0.2037 - val_acc: 0.8260
Epoch 66/100
7429/7429 [==============================] - 0s 49us/step - loss: 0.2156 - acc: 0.8203 - val_loss: 0.2036 - val_acc: 0.8260
Epoch 67/100
7429/7429 [==============================

In conclusion, from the previous code we were able to see that the loss started off extremely high and depreciate insignificantly to allievate this issue I map the output to 0 and 1 then re run the model again. We get a more desirable range of result.

In [12]:
#Evaluating on the test set
model.evaluate(X_test, Y_test)[1]#Using index 1 because the function returns the loss as the first element and the accuracy as the second element.

1592/1592 [==============================] - 0s 38us/step


0.8316582914572864

In [13]:
#Test the model on the existing data set that have been split into train and test
prediction = model.predict(X_test,verbose=1)
prediction = scaler.inverse_transform(predictions)
prediction_df=pd.DataFrame(prediction)
prediction_df.to_csv('new.csv')

#Output X_test to verify the result of the output
X_testdf = pd.DataFrame(X_test)
X_testdf.to_csv('new1.csv')

1592/1592 [==============================] - 0s 41us/step


NameError: name 'predictions' is not defined

## Visualizing Loss And Accuracy

In [ ]:
import matplotlib.pyplot as plt

plt.plot(hist.history['loss'])#plotting the loss
plt.plot(hist.history['val_loss'])#plotting the val_loss
plt.title('Model loss')#Title of the graph
plt.ylabel('Loss')
plt.xlabel('Epoch')
plt.legend(['Train', 'Val'], loc='upper right')
plt.show()

In [ ]:
plt.plot(hist.history['acc'])
plt.plot(hist.history['val_acc'])
plt.title('Model accuracy')
plt.ylabel('Accuracy')
plt.xlabel('Epoch')
plt.legend(['Train', 'Val'], loc='lower right')
plt.show()

## Adding Regularization to the Neural Network

In [ ]:
model_2 = Sequential([
    Dense(1000, activation='relu', input_shape=(3,)),
    Dense(1000, activation='relu'),
    Dense(1000, activation='relu'),
    Dense(1000, activation='relu'),
    Dense(1, activation='sigmoid'),
])

model_2.compile(optimizer='adam',
               loss='binary_crossentropy',
               metrics=['accuracy'])

hist_2 = model_2.fit(X_train, Y_train,
                    batch_size=32, epochs=50,
                    validation_data = (X_val, Y_val))

In [ ]:
plt.plot(hist_2.history['loss'])
plt.plot(hist_2.history['val_loss'])
plt.title('Model loss')
plt.ylabel('Loss')
plt.xlabel('Epoch')
plt.legend(['Train', 'Val'], loc='upper right')
plt.show()

The training loss is decreasing, but the validation loss is way above the training loss and increasing hence we can see that this is the clear sign of the over-fitting.

In [ ]:
#plotting the accuracy code
plt.plot(hist_2.history['acc'])
plt.plot(hist_2.history['val_acc'])
plt.title('Model accuracy')
plt.ylabel('Accuracy')
plt.xlabel('Epoch')
plt.legend(['Train', 'Val'], loc='lower right')
plt.show()

In [ ]:
#Removing over-fitting with L2 regularization and dropout
from keras.layers import Dropout
from keras import regularizers

#kernel_regularizer tells Keras to include the squared values of those parameters

#our overall loss function, and weight them by 0.01 in the loss function
model_3 = Sequential([
    Dense(1000, activation='relu', kernel_regularizer=regularizers.l2(0.01), input_shape=(3,)),
    Dropout(0.3),
    Dense(1000, activation='relu', kernel_regularizer=regularizers.l2(0.01)),
    Dropout(0.3),
    Dense(1000, activation='relu', kernel_regularizer=regularizers.l2(0.01)),
    Dropout(0.3),
    Dense(1000, activation='relu', kernel_regularizer=regularizers.l2(0.01)),
    Dropout(0.3),
    Dense(1, activation='sigmoid', kernel_regularizer=regularizers.l2(0.01)),
])

model_3.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])
hist_3 = model_3.fit(X_train, Y_train,
          batch_size=32, epochs=50,
          validation_data=(X_val, Y_val))

In [ ]:
#Visualising validation and training loss
plt.plot(hist_3.history['loss'])
plt.plot(hist_3.history['val_loss'])
plt.title('Model loss')
plt.ylabel('Loss')
plt.xlabel('Epoch')
plt.legend(['Train', 'Val'], loc='upper right')
plt.ylim(top=1.2, bottom=0)
plt.show()


In [ ]:
#Plotting the accuracy
plt.plot(hist_3.history['acc'])
plt.plot(hist_3.history['val_acc'])
plt.title('Model accuracy')
plt.ylabel('Accuracy')
plt.xlabel('Epoch')
plt.legend(['Train', 'Val'], loc='lower right')
plt.show()